GB+Tree rebalance and merge

burqen opened the pull request:
**What?**
Implementation of rebalance and merge.

**Why?**
When removing keys from GB+Tree we want to avoid having leaves with key count less than half of max key count. This is done by rebalancing keys among sibling nodes if they between themselves have enough to at least get `maxKeyCount/2` keys each. Otherwise we will merge them together, effectively deleting the node that we merge from.

**Problems?**
Seek in GB+Tree rely on no keys ever being moved from right to left in the tree. Therefore we are not allowed to rebalance or merge from right to left. This creates a somewhat asymmetric implementation.

Commit:
GBPTree - Rebalance support: Can insert multiple empty slots in tree-node

Commit:
GBPTree - Rebalance support: Rename fields in StructurePropagation

Commit:
GBPTree - Rebalance support: Small cleanup in InternalTreeLogic

Commit:
GBPTree - Rebalance implementation

When removing keys from a leaf an underflow can occur. If left sibling
and current node combined has more keys than what can fit in a single node
a rebalance occurs.

A rebalance move keys from left sibling to current node to even out the number of
keys between them. This makes the keys more evenly distributed among the leaves.

Because we change how the keys are split among the two nodes the keys splitting
the ranges further up in the tree need to be replaced. This is communicated through
StructurePropagation.

A new version of left sibling and/or current node is created as necessary and this
is also communicated up the tree trough StructurePropagation.

Commit:
GBPTree - Rebalance test: InternalTreeLogicTest clean up

Create separate read cursor that can be freely be moved around when reading
from the tree to avoid moving cursro used by InternalTreeLogic.

Commit:
GBPTree - Rebalance test: Unit tests in InternalTreeLogicTest

Commit:
GBPTree - Rebalance test: Unit tests in SeekCursorTest

Commit:
GBPTree - Better exception messages from RightmostInChain

and more explicit variable names

Commit:
GBPTreeIT use check-pointing and avoid performing consistency check in @After

Checkpointing exposes a lot of new code paths to be executed and should naturally
be part of concurrency testing.

If consistency check failed in @After, no random seed would be printed.

Commit:
GBPTree - Merge implemented

An underflow in some node has occured and we want to merge
node L into node R. L is left sibling to R and vice versa.
It is done like this:

Phase 1 (Move data)
1. Left sibling of L is LL. Set right sibling of LL to be R and
   left sibling of R to be LL. Effectively removing L from the leaf chain.
2. Set new generation of L to be R.
3. Add L to freelist.
4. Copy all keys and values from L to R.
5. Update partent pointers to match changes, new pointer to both L and R.

Phase 2 (Update splitter key in internal nodes)
   Parent of L and R (potentially same internal node) now both hold pointer
to the same leaf, R. This is ok for now, because it just means we have to
different paths to the same leaf. We do however want to remove one of those paths
together with some key to not store more than what we need among the
internal nodes.
   L and R are part of different subtrees, subL and subR.
(Note that it is possible that subL and subR only contain L och R respectively.)
Because L and R are sibling, we know that subL and subR are splitted by exactly
one key, the 'splitter', which is located in the closest common parent of
subL and subR. subL is located just to the left of splitter and subR just to the right.
splitter should be replaced by the key that previously would guide a search towards L.
This key is the rightmost key in subL. If no such key can be found, if subL only
contains L or if none of the internal nodes in subL has any keys, then splitter
is removed together with subL.
   So:
6. Bubble up rightmost key from subL.
7. Every internal node on the bubble path that contains zero keys can be
   added to freelist.
8. Replace splitter with bubble key or remove if no bubble key could be found.
9. If root end up with zero keys, shrink the tree.

Commit:
GBPTree - Merge: InternalTreeLogic unit testing

and some cleanup

Commit:
GBPTree - Merge: SeekCursor unit testing

and some cleanup

Commit:
Fix checkstyle

tinwelint commented on community/index/src/main/java/org/neo4j/index/internal/gbptree/InternalTreeLogic.java:
```diff
@@ -849,11 +854,55 @@ VALUE remove( PageCursor cursor, StructurePropagation<KEY> structurePropagation,
+        // Left sibling is not allowed to be NO_NODE here because that means there is a child node with no parent
+        PointerChecking.checkPointer( leftSibling, false );
+
+        bTreeNode.goTo( cursor, "left sibling", leftSibling );
```
In other places where we temporarily go to another node and then back we usually use a linked cursor, do that here too?

tinwelint commented on community/index/src/test/java/org/neo4j/index/internal/gbptree/GBPTreeIT.java:
```diff
@@ -77,78 +76,80 @@
-    public void consistencyCheckAndClose() throws IOException
-    {
-        threadPool.shutdownNow();
-        index.consistencyCheck();
```
The need to merge this "After" code into the test seems to be a bug in RandomRule, doesn't it? It's fine, but it should be fixed so that tests generally don't need to go through this process too.

tinwelint reviewed (approved):
Looks great. Much of the code is documented or self-documenting in a way. I did some QA of alternating inserting,removing,checkpoint and can see that file size stays flat.

tinwelint committed:
Uses linked cursor for updating siblings

tinwelint committed:
Puts consistency check back in @After clause

next time issue with RandomRule not printing seed comes up, it should be fixed instead.
